File: src/main/java/nl/hanze/week1/assignment1/StockGrabber.java
package nl.hanze.week1.assignment1;

import java.util.ArrayList;
import java.util.List;

public class StockGrabber implements Subject {
    private List<Observer> observers = new ArrayList<>();
    private double ibmPrice;
    private double aaplPrice;
    private double googPrice;

    public void register(Observer newObserver) {
        observers.add(newObserver);
    }

    public void unregister(Observer newObserver) {
        observers.remove(newObserver);
    }

    public void notifyObservers() {

        for (Observer observer : observers) {
            observer.update(ibmPrice, aaplPrice, googPrice);
        }
    }

    public void setIBMPrice(double newIBMPrice) {

        this.ibmPrice = newIBMPrice;
        notifyObservers();

    }

    public void setAAPLPrice(double newAAPLPrice) {

        this.aaplPrice = newAAPLPrice;
        notifyObservers();
    }

    public void setGOOGPrice(double newGOOGPrice) {

        this.googPrice = newGOOGPrice;
        notifyObservers();
    }
}

File: src/main/java/nl/hanze/assignment1/StockGrabber.kt
package nl.hanze.assignment1

import nl.hanze.assignment1.observers.Observer
import kotlin.properties.Delegates.observable

class StockGrabber : Subject {
    private val observers = mutableListOf<Observer>()

    var ibmPrice by observable(0.0){_,_,new -> notifyObservers(Stock.IBM updateTo new) }
    var aaplPrice by observable(0.0){_,_,new -> notifyObservers(Stock.AAPL updateTo new) }
    var googPrice by observable(0.0){ _, _, new -> notifyObservers(Stock.GOOGLE updateTo new) }

    override fun register(observer: Observer) {
        observers+= observer
    }

    override fun unregister(observer: Observer) {
        observers+= observer
    }

    override fun notifyObservers(stockUpdate: StockUpdate)
            = observers.forEach { it.update(stockUpdate) }
}
